### PR TITLE
Prefs [TASK][28596] Create presenter and contract restaurant tab

### PR DESCRIPTION
**Pull request by viet06061999**
## Related Tickets
- [Task#28596: Create presenter and contract restaurant tab](https://edu-redmine.sun-asterisk.vn/issues/28596)

## WHAT
Create Presenter and contract restaurant tab
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/restaurant/RestaurantContract.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    interface View : BaseContract.View<Presenter> {
+        fun updateListLocation(list: List<Location>)
+        fun navigationToDetail(meal: String? = null, restaurant: String? = null, item: Location)
+    }
+
+    interface Presenter : BaseContract.Presenter<View> {
+        fun onQuerySearch(query: String)
+        fun onItemClick(meal: String? = null, restaurant: String? = null, item: Location)
+    }
+}
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/restaurant/RestaurantContract.kt]: những cái logic chỉ liên quan tới View như này thì để trên View thôi, chuyển qua presenter làm phức tạo vấn đề hơn

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/restaurant/RestaurantContract.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import com.sunasterisk.travelapp.data.models.Location
+import com.sunasterisk.travelapp.ui.base.BaseContract
+
+interface RestaurantContract {
+    interface View : BaseContract.View<Presenter> {
+        fun updateListLocation(list: List<Location>)
+    }
+
+    interface Presenter : BaseContract.Presenter<View> {
+        fun onQuerySearch(query: String)
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/restaurant/RestaurantContract.kt]:
```suggestion
        fun searchRestaurant(input: String)
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/restaurant/RestaurantContract.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.sunasterisk.travelapp.ui.home.restaurant
+
+import com.sunasterisk.travelapp.data.OnDataCallback
+import com.sunasterisk.travelapp.data.models.Location
+import com.sunasterisk.travelapp.ui.base.BaseContract
+
+interface RestaurantContract {
+    interface View : BaseContract.View<Presenter> {
+        fun updateListLocation(list: List<Location>)
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/restaurant/RestaurantContract.kt]:
```suggestion
        fun updateLocations(list: List<Location>)
```

**Review by aug16vcc — APPROVED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPFragment.kt [hunk 4]
```diff
@@ -1,18 +1,28 @@
...
+    BaseFragment(),
+    BaseContract.View<T> {
+
+    protected abstract val presenter: T
     private var progressDialog: Dialog? = null
 
+    @Suppress("UNCHECKED_CAST")
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        presenter.onAttach(this as V)
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPFragment.kt]: check cast trước khi cast chứ 
```suggestion
        if(this is V){
            presenter.onAttack(this)
        }
```

Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPFragment.kt]: vâng ạ, Nhưng V là loại General nên ko check cast được ạ

Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPFragment.kt]: thế thì kiểu cast nullable có được k 
```suggestion
        (this as? V)?.let { presenter.onAttach(it) }
```

Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPFragment.kt]: vâng được ạ

**Pull request merged**